SSA/Experimental/Bits/Fast/Tactic.lean

```diff
@@ -1,491 +0,0 @@
-      }
-    | _ => throwError m!"Expression {e} is not of the expected form. Expected something of the form BitStream.EqualUpTo (w : Nat) (lhs : BitStream) (rhs : BitStream) : Prop"
-
-/--
-Introduce vars which maps variable ids to the variable values.
-
-let vars (n : Nat) : BitStream := BitStream.ofBitVec (if n = 0 then v0 else if n = 1 then v1 else if n = 2 then v2 ......)
-
-Term.var 0 -- represent the 0th variable
-Term.var 1 -- represent the 1st variable
--/
-
-def introduceMapIndexToFVar : TacticM Unit := withMainContext <|  do
-  let context : LocalContext ← getLCtx
-  let fVars : List FVarId :=  (PersistentArray.toList context.decls).filterMap (fun d => match d with
-    | .none => .none
-    | .some (.cdecl _ f _ type _ _) =>  match (type : Q(Type)) with
-      | .app (.const ``BitVec []) _ => .some f
-      | _ => .none
-    | .some (.ldecl _ _ _ _ _ _ _) => .none
-    )
-  let goal : MVarId ← getMainGoal
-  let last : FVarId := fVars.get! 0
-  let mapIndexToFVarType: Q(Type) := q(Nat → BitStream)
-  let lastBVar : Q(Nat) := .bvar 0
-  let target : Expr ← getMainTarget
-  match_expr target  with
-    | BitStream.EqualUpTo a _ _  => do
-      let length : Expr := a
-      let hypValue : Expr  := fVars.foldl (fun (accumulator : Expr) (currentFVar : FVarId) =>
-        let quotedCurrentFVar : Expr := .fvar currentFVar
-        let fVarId : Q(Nat) := quoteFVar currentFVar
-        let eqE : Q(Prop)  := q($lastBVar = $fVarId);
-        ((((((Expr.const `ite [Level.zero.succ]).app (.app (.const ``BitVec []) length)).app
-                        eqE).app
-                    (((Expr.const ``instDecidableEqNat []).app lastBVar).app fVarId)).app
-                quotedCurrentFVar).app
-            accumulator)
-        ) (.fvar last)
-      let mapIndexToFVar : Q(Nat → BitStream):=
-        (Expr.lam `n (Expr.const `Nat [])
-          (((Expr.const `BitStream.ofBitVec []).app
-                length).app
-            hypValue)
-          BinderInfo.default)
-      let newGoal : MVarId ← goal.define `vars mapIndexToFVarType mapIndexToFVar
-      replaceMainGoal [newGoal]
-    | _ => throwError "Goal is not of the expected form"
-
-elab "introduceMapIndexToFVar" : tactic => introduceMapIndexToFVar
-
-/--
-Create bv_automata tactic which solves equalities on bitvectors.
--/
-
-macro "bv_automata" : tactic =>
-  `(tactic| (
-  apply BitStream.eq_of_ofBitVec_eq
-  introduceMapIndexToFVar
-  intro mapIndexToFVar
-  simp only [
-    reduce_bitvec2,
-  ]
-  try simp only [
-    ← eval_sub,
-    ← eval_add,
-    ← eval_neg,
-    ← eval_and,
-    ← eval_xor,
-    ← eval_or,
-    ← eval_not,
-    Nat.reduceAdd,
-    BitVec.ofNat_eq_ofNat
-  ]
-  intros _ _
-  apply congrFun
-  apply congrFun
-  native_decide
-  ))
-
-/-!
-# Test Cases
--/
-
-def alive_1 {w : ℕ} (x x_1 x_2 : BitVec w) : (x_2 &&& x_1 ^^^ x_1) + 1#w + x = x - (x_2 ||| ~~~x_1) := by
-  bv_automata
```

Comment:
These test cases might be worth reviving whenever your tactic is powerful enough.

Reply:
Done, I've revived them now.